src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -191,224 +204,233 @@ private void CreateStreamAndStartWatcher()
-            }
-            else
-            {
-                // We failed and need to release the caller so the OnError event can be processed
-                _runLoopStartedEvent.Set();
+                    // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                    Interop.RunLoop.CFRunLoopRun();
 
-                // An error occurred while trying to start the run loop so fail out
-                Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
-                OnError(new ErrorEventArgs(new IOException(SR.EventStream_FailedToStart, Marshal.GetLastWin32Error())));
-            }
-        }
+                    // When we get here, we've requested to stop so cleanup the EventStream and unschedule from the RunLoop
+                    Interop.EventStream.FSEventStreamStop(_eventStream);
+                }
+                else
+                {
+                    // We failed and need to release the caller so the OnError event can be processed
+                    _runLoopStartedEvent.Set();
 
-        private void FileSystemEventCallback( 
-            FSEventStreamRef streamRef, 
-            IntPtr clientCallBackInfo, 
-            size_t numEvents, 
-            String[] eventPaths, 
-            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
-            Interop.EventStream.FSEventStreamEventFlags[] eventFlags,
-            [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)]
-            FSEventStreamEventId[] eventIds)
-        {
-            Debug.Assert((numEvents.ToInt32() == eventPaths.Length) && (numEvents.ToInt32() == eventFlags.Length) && (numEvents.ToInt32() == eventIds.Length));
+                    // Try to get the Watcher to raise the error event; if we can't do that, just silently exist since the watcher is gone anyway
+                    FileSystemWatcher watcher;
+                    if (_weakWatcher.TryGetTarget(out watcher))
+                    {
+                        // An error occurred while trying to start the run loop so fail out
+                        watcher.OnError(new ErrorEventArgs(new IOException(SR.EventStream_FailedToStart, Marshal.GetLastWin32Error())));
+                    }
+                }
 
-            // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-            // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-            List<FSEventStreamEventId> handledRenameEvents = null;
+                // Always unschedule the RunLoop before cleaning up
+                Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
```

Comment:
It looks unlikely but possible that an exception could come out of the OnError call, if for example it invoked the user's error event handler synchronously and that event handler threw.  Should this be in a finally block?